src/rabbit_channel_interceptor.erl

```diff
@@ -14,73 +14,105 @@
-
--callback intercept(original_method(), rabbit_types:vhost()) ->
-    processed_method() | rabbit_misc:channel_or_connection_exit().
-
-%% Whether the interceptor wishes to intercept the amqp method
--callback applies_to(intercept_method()) -> boolean().
+%% Derive some initial state from the channel. This will be passed back
+%% as the third argument of intercept/3.
+-callback init(rabbit_channel:channel()) -> any.
+-callback intercept(original_method(), original_content(), any) ->
+    {processed_method(), processed_content()} | rabbit_misc:channel_or_connection_exit().
+-callback applies_to() -> list(method_name()).
 
 -else.
 
 -export([behaviour_info/1]).
 
 behaviour_info(callbacks) ->
-    [{description, 0}, {intercept, 2}, {applies_to, 1}];
+    [{description, 0}, {init, 1}, {intercept, 3}, {applies_to, 0}];
 behaviour_info(_Other) ->
     undefined.
 
 -endif.
 
-%%----------------------------------------------------------------------------
-
-intercept_method(#'basic.publish'{} = M, _VHost) -> M;
-intercept_method(#'basic.ack'{}     = M, _VHost) -> M;
-intercept_method(#'basic.nack'{}    = M, _VHost) -> M;
-intercept_method(#'basic.reject'{}  = M, _VHost) -> M;
-intercept_method(#'basic.credit'{}  = M, _VHost) -> M;
-intercept_method(M, VHost) ->
-    intercept_method(M, VHost, select(rabbit_misc:method_record_type(M))).
-
-intercept_method(M, _VHost, []) ->
-    M;
-intercept_method(M, VHost, [I]) ->
-    M2 = I:intercept(M, VHost),
-    case validate_method(M, M2) of
-        true ->
-            M2;
-        _   ->
-            internal_error("Interceptor: ~p expected "
-                                "to return method: ~p but returned: ~p",
-                                [I, rabbit_misc:method_record_type(M),
-                                 rabbit_misc:method_record_type(M2)])
-    end;
-intercept_method(M, _VHost, Is) ->
-    internal_error("More than one interceptor for method: ~p -- ~p",
-                   [rabbit_misc:method_record_type(M), Is]).
-
-%% select the interceptors that apply to intercept_method().
-select(Method)  ->
-    [M || {_, M} <- rabbit_registry:lookup_all(channel_interceptor),
-          code:which(M) =/= non_existing,
-          M:applies_to(Method)].
+init(Ch) ->
+    Mods = [M || {_, M} <- rabbit_registry:lookup_all(channel_interceptor)],
+    check_no_overlap(Mods),
+    to_fn(Mods, Ch).
+
+%% Turn a list of interceptor modules into a single function
+to_fn(Mods, Ch) ->
+    case Mods of
+        [] -> fun(M, C) -> {M, C} end;
+        [Mod] -> to_fn1(Mod, Ch);
+        [Mod|Mods1] ->
+            Fn1 = to_fn1(Mod, Ch),
+            Fn2 = to_fn(Mods1, Ch),
+            fun(M, C) ->
+                {M1, C1} = Fn1(M, C),
+                Fn2(M1, C1)
+            end
+    end.
+
+%% Turn a single interceptor module into a function
+to_fn1(Mod, Ch) ->
+    St = Mod:init(Ch),
+    fun (M, C) ->
+        % this little dance is because Mod might be unloaded at any point
```

Comment:
So instead of refreshing the state cache we just handle the module going away right? I think it's fine since channels might not be so long-lived to have this be a frequent issue.